### ISSUE TITLE
Block/Unblock user

### ISSUE THREAD

**Issue by Shahjahan786** [QuickBlox/quickblox-android-sdk#580]
I have trouble in blocking/unblocking user:
when I block some user, I'm unable to send messages to all other users also,
if I Unblock the user still I'm unable to send messages to all users.

I'm using following method to block/unblock a user Kindly suggest changes.

    public static boolean blockUnBlockUser(int userId, boolean allow) {

        boolean success = false;
        QBPrivacyList qbPrivacyList = null;
        QBPrivacyListsManager qbPrivacyListsManager = QBChatService.getInstance().getPrivacyListsManager();

        if (qbPrivacyListsManager == null) {
            AppUtils.logDebug("qbPrivacyListsManager", "qbPrivacyListsManager is null");
            return false;
        }

        try {
            qbPrivacyList = qbPrivacyListsManager.getPrivacyList(PRIVACY_LIST_NAME);
        } catch (Exception e) {
            e.printStackTrace();
        }

        if (qbPrivacyList == null) { // CREATE NEW PRIVACY LIST
            qbPrivacyList = new QBPrivacyList();
            qbPrivacyList.setName(PRIVACY_LIST_NAME);
        }

        QBPrivacyListItem qbPrivacyListItem = new QBPrivacyListItem();
        qbPrivacyListItem.setType(QBPrivacyListItem.Type.USER_ID);
        qbPrivacyListItem.setValueForType(String.valueOf(userId));

        ArrayList<QBPrivacyListItem> qbPrivacyListItems = new ArrayList<>();

        if (null != qbPrivacyList.getItems()) {
            qbPrivacyListItems.addAll(qbPrivacyList.getItems());
        }

        boolean itemFound = false;
        for (int i = 0; i < qbPrivacyListItems.size(); i++) {
            QBPrivacyListItem listItem = qbPrivacyListItems.get(i);

            if (listItem.getValueForType().equals(qbPrivacyListItem.getValueForType())) {

                qbPrivacyListItem.setAllow(allow);
                qbPrivacyListItem.setMutualBlock(!allow);
                qbPrivacyListItems.add(qbPrivacyListItem);

                itemFound = true;
                break;
            }
        }

        if (!itemFound && !allow) {
            qbPrivacyListItem.setAllow(false);
            qbPrivacyListItem.setMutualBlock(true);
            qbPrivacyListItems.add(qbPrivacyListItem);
        }

        qbPrivacyList.setItems(qbPrivacyListItems);

        try {
            qbPrivacyListsManager.declinePrivacyList();
            qbPrivacyListsManager.createPrivacyList(qbPrivacyList);
            qbPrivacyListsManager.applyPrivacyList(PRIVACY_LIST_NAME);
            AppUtils.logDebug("PRIVACY_LIST[ITEMS]", qbPrivacyList.getItems() +"");
            success = true;
        } catch (Exception e) {
            success = false;
            e.printStackTrace();
        }
        return success;
    }

**Comment by Shahjahan786** [comment]
Anyone to help on this?

**Comment by RomanPronin** [comment]
Hi @Shahjahan786, you should update an item in your qbPrivacyListItems, not add the same one, look: 
```
        for (int i = 0; i < qbPrivacyListItems.size(); i++) {
            QBPrivacyListItem listItem = qbPrivacyListItems.get(i);

            if (listItem.getValueForType().equals(qbPrivacyListItem.getValueForType())) {

                qbPrivacyListItem.setAllow(allow);
                qbPrivacyListItem.setMutualBlock(!allow);
//                qbPrivacyListItems.add(qbPrivacyListItem);
                qbPrivacyListItems.set(i, qbPrivacyListItem);

                itemFound = true;
                break;
            }
        }

**Comment by Shahjahan786** [comment]
Thanks @RomanPronin After making suggested changes block/unblock functionality looks working, can you please suggest any other code changes/optimisations for the `lockUnBlockUser(int userId, boolean allow)` method.

**Comment by RomanPronin** [comment]
Your function looks fine. All I can add it's a couple convenient methods:
```
    private QBPrivacyListItem createPrivacylistItem(QBPrivacyListItem.Type type, String value) {

        QBPrivacyListItem qbPrivacyListItem = new QBPrivacyListItem();
        qbPrivacyListItem.setType(type);
        qbPrivacyListItem.setValueForType(value);
        qbPrivacyListItem.setMutualBlock(true);
        return qbPrivacyListItem;
    }

    private QBPrivacyList createPrivacylist(String name, QBPrivacyListItem... privacyListItems) throws Exception {
        //create record for privacy list
        ArrayList<QBPrivacyListItem> items = new ArrayList<>();

        Collections.addAll(items, privacyListItems);

        QBPrivacyListsManager privacyListsManager = chatService.getPrivacyListsManager();
        QBPrivacyList privacyList = new QBPrivacyList();
        privacyList.setName(name);
        privacyList.setItems(items);
        privacyListsManager.createPrivacyList(privacyList);

        return privacyList;
    }
// and use it
        QBPrivacyListItem privacyListItem = createPrivacylistItem(QBPrivacyListItem.Type.USER_ID, String.valueOf(PARTICIPANT_ID));
        QBPrivacyList privacyList = createPrivacylist(USER_PRIVACY_LIST_NAME, privacyListItem);
        chatService.getPrivacyListsManager().applyPrivacyList(privacyList);

**Comment by QB-nikolay-zolotarov** [comment]
Hello QuickBlox customer,

This is Nikolay from QuickBlox support.

The issue was closed as it was outdated.

Please check the relevant section of our documentation here:
[https://docs.quickblox.com/docs/android-chat-privacy-list](https://docs.quickblox.com/docs/android-chat-privacy-list)

Also, please update the SDK to the latest version:
[https://github.com/QuickBlox/quickblox-android-sdk-releases/releases/tag/3.9.11](https://github.com/QuickBlox/quickblox-android-sdk-releases/releases/tag/3.9.11)

If it is still relevant after reviewing the updated information, feel free to open a new issue.

Have a nice day.
